src/feeds/FeedReader.ts

```diff
@@ -0,0 +1,226 @@
+                } else {
+                    throw err;
+                }
+            });
+            if (!validateAccountData(accountData)) {
+                const errors = validateAccountData.errors!.map(e => `${e.instancePath} ${e.message}`);
+                throw new Error(`Invalid account data: ${errors.join(', ')}`);
+            }
+            for (const url in accountData) {
+                this.seenEntries.set(url, accountData[url]);
+            }
+        } catch (err: unknown) {
+            log.error(`Failed to load seen feed entries from accountData: ${err}. This may result in skipped entries`);
+            // no need to wipe it manually, next saveSeenEntries() will make it right
+        }
+    }
+
+    private async saveSeenEntries(): Promise<void> {
+        const accountData: AccountData = {};
+        for (const [url, guids] of this.seenEntries.entries()) {
+            accountData[url.toString()] = guids;
+        }
+        await this.matrixClient.setAccountData(FeedReader.seenEntriesEventType, accountData);
+    }
+
+    private async pollFeeds(): Promise<void> {
+        log.debug(`Checking for updates in ${this.observedFeedUrls.size} RSS/Atom feeds`);
+
+        let seenEntriesChanged = false;
+
+        const fetchingStarted = (new Date()).getTime();
+
+        for (const url of this.observedFeedUrls.values()) {
+            try {
+                const res = await axios.get(url.toString());
+                const feed = await (new Parser()).parseString(res.data);
+                let initialSync = false;
+                let seenGuids = this.seenEntries.get(url);
+                if (!seenGuids) {
+                    initialSync = true;
+                    seenGuids = [];
+                    seenEntriesChanged = true; // to ensure we only treat it as an initialSync once
+                }
+                const seenGuidsSet = new Set(seenGuids);
+                const newGuids = [];
+                log.debug(`Found ${feed.items.length} entries in ${url}`);
+                for (const item of feed.items) {
+                    const guid = item.guid || item.id || item.link || item.title;
+                    if (!guid) {
+                        log.error(`Could not determine guid for entry in ${url}, skipping`);
+                        continue;
+                    }
+                    newGuids.push(guid);
+
+                    if (initialSync) {
+                        log.debug(`Skipping entry ${guid} since we're performing an initial sync`);
+                        continue;
+                    }
+                    if (seenGuidsSet.has(guid)) {
+                        log.debug('Skipping already seen entry', guid);
+                        continue;
+                    }
+
+                    const entry = {
+                        feed: { title: stripHtml(feed.title!), url: url.toString() },
+                        title: stripHtml(item.title!),
+                        link: item.link!,
```

Comment:
Presumably any of these could be undefined right? Do we need to be careful here and check?

Reply:
Could happen, yeah. I made sure we're making sense of the edgecases in b8afd4f7ef5248054c1deb32286f0275612935f2